src/wallet/walletdb.cpp

```diff
@@ -943,6 +1080,100 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         }
     }
 
+    // Upgrade to using KeyMan
+    // Find the keys which are used in single key internal and external descriptors with
+    // the pre-taproot output types
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS) &&
+        !pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) &&
+        (!pwallet->IsWalletFlagSet(WALLET_FLAG_USES_KEYMAN) ||
+        last_client <= 229999)
+        ) {
+        std::map<CExtPubKey, std::pair<std::map<std::string, int>, uint64_t>> descs_keys;
+        std::map<std::string, int> tmpl = {{"pkh(", 0}, {"sh(wpkh(", 0}, {"wpkh(", 0}};
+
+        // Find root xpubs used in pkh(), sh(wpkh()), and wpkh() descriptors
+        for (const auto& spkm : pwallet->GetAllScriptPubKeyMans()) {
+            const DescriptorScriptPubKeyMan* desc_spkm = dynamic_cast<DescriptorScriptPubKeyMan*>(spkm);
+            assert(desc_spkm);
+            LOCK(desc_spkm->cs_desc_man);
+            WalletDescriptor w_desc = desc_spkm->GetWalletDescriptor();
+
+            if (!w_desc.descriptor->IsRange()) {
+                continue;
+            }
+
+            std::set<CPubKey> desc_pubkeys;
+            std::set<CExtPubKey> desc_xpubs;
+            w_desc.descriptor->GetPubkeys(desc_pubkeys, desc_xpubs);
+            if (desc_xpubs.size() != 1) {
+                continue;
+            }
+            const CExtPubKey& xpub = *desc_xpubs.begin();
+
+            if (descs_keys.count(xpub) == 0) {
+                descs_keys.emplace(xpub, std::make_pair(tmpl, 0));
+            }
+
+            const std::string desc = w_desc.descriptor->ToString();
+            if (desc.find("pkh(") == 0) {
```

Comment:
nit:
```suggestion
            if (desc.find("pkh(")) {
```

Reply:
Done

Reply:
Are you sure? `find` returns the position, so this should match `wpkh(…)` but not `pkh()`. The `else if`'s seem wrong for the same reason.

Reply:
Maybe `desc.rfind("pkh(", 0) == 0` ?

Reply:
I think the original `desc.find("pkh(") == 0` was fine.

Reply:
Ah, yes. Looked at the docs this time. I've changed these all to be `== 0`.